pop-api/src/error.rs

```diff
@@ -0,0 +1,255 @@
+use crate::assets::use_cases::fungibles::{convert_to_fungibles_error, FungiblesError};
+use ink::env::chain_extension::FromStatusCode;
+use scale::{Decode, Encode};
+use PopApiError::*;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+#[repr(u8)]
+pub enum PopApiError {
+	/// Some unknown error occurred. Go to the Pop API docs section `Pop API error`.
+	Other {
+		// Index within the `DispatchError`
+		dispatch_error_index: u8,
+		// Index within the `DispatchError` variant.
+		error_index: u8,
+		// Index for further nesting, e.g. pallet error.
+		error: u8,
+	} = 0,
+	/// Failed to lookup some data.
+	CannotLookup = 1,
+	/// A bad origin.
+	BadOrigin = 2,
+	/// A custom error in a module.
+	Module {
+		index: u8,
+		error: u8,
+	} = 3,
+	/// At least one consumer is remaining so the account cannot be destroyed.
+	ConsumerRemaining = 4,
+	/// There are no providers so the account cannot be created.
+	NoProviders = 5,
+	/// There are too many consumers so the account cannot be created.
+	TooManyConsumers = 6,
+	/// An error to do with tokens.
+	Token(TokenError) = 7,
+	/// An arithmetic error.
+	Arithmetic(ArithmeticError) = 8,
+	/// The number of transactional layers has been reached, or we are not in a transactional
+	/// layer.
+	Transactional(TransactionalError) = 9,
+	/// Resources exhausted, e.g. attempt to read/write data which is too large to manipulate.
+	Exhausted = 10,
+	/// The state is corrupt; this is generally not going to fix itself.
+	Corruption = 11,
+	/// Some resource (e.g. a preimage) is unavailable right now. This might fix itself later.
+	Unavailable = 12,
+	/// Root origin is not allowed.
+	RootNotAllowed = 13,
+	// TODO: make generic and add docs.
+	UseCaseError(FungiblesError) = 254,
+	DecodingFailed = 255,
+}
+
+impl FromStatusCode for PopApiError {
+	fn from_status_code(status_code: u32) -> core::result::Result<(), Self> {
+		match status_code {
+			0 => Ok(()),
+			_ => Err(convert_to_pop_api_error(status_code)),
```

Comment:
Would prefer `status_code.into()` and then implementing a trait which converts from a u32 into the error. That is usable anywhere in the crate rather than needing to know about `convert_to_pop_api_error`

Reply:
Outdated, see refactor